### ISSUE TITLE
Hid function not working

### ISSUE THREAD

**Issue by sploitpayt** [ace2nutzer/Samsung_dreamlte_Kernel#2]
Hid fuction not working even after enabled it and the device not generating hidg0 and hidg1 file in /dev/

**Comment by ace2nutzer** [comment]
What did you enabled ? - CONFIG_???

**Comment by sploitpayt** [comment]
device drivers -> usb support -> usb gadget support -> hid function=y (menuconfig)

CONFIG_USB_F_HID=y

**Comment by ace2nutzer** [comment]
try to enable also
USB_G_HID

**Comment by sploitpayt** [comment]
got this error https://del.dog/mihanylast.txt when enabled [ #CONFIG_USB_G_HID is not set ] -> [ CONFIG_USB_G_HID=y ]

**Comment by ace2nutzer** [comment]
Add
`#include <linux/if_ether.h>	/* ETH_ALEN */`
[here](https://github.com/ace2nutzer/Samsung_dreamlte_Kernel/blob/f5438072c97c9ad56d28f30ccda20681d5a6b6b2/drivers/usb/gadget/configfs.c#L24)

**Comment by sploitpayt** [comment]
As per you said i added that on line 24 in 'configfs.c' and got the same error again https://del.dog/mihanylast.txt

**Comment by ace2nutzer** [comment]
Strange .. it should work. Maybe later i'll try by myself. Honestly this is very easy to fix since you have only to provide the proper missing header file.

**Comment by sploitpayt** [comment]
Ok thx and hid function is a common feature most device with kernel above 4.x has enabled it by default

**Comment by ace2nutzer** [comment]
In stock defconfig it's disabled, and so far nobody asked for it.
Later i'll fix it, then you only need to fetch your repo and of course enable USB_G_HID.

**Comment by sploitpayt** [comment]
ok

**Comment by ace2nutzer** [comment]
btw USB_G_HID is probably not needed.
The right config is:
```
CONFIG_USB_CONFIGFS=y
CONFIG_USB_CONFIGFS_F_HID=y
CONFIG_USB_F_HID=y
```
But gives also a build error ... did you tried it ?

**Comment by sploitpayt** [comment]
Littile confused by your words u mean in s8 we cannot have hidg0 in /dev because of dual-role port

**Comment by ace2nutzer** [comment]
Yes, because it's not a normal USB port like that from a PC. But the devices will work anyway normal with the proper configfs.
I guess you wanted to connect a USB-keyboard or USB-mouse, right ?

**Comment by sploitpayt** [comment]
my goal is to send key strokes from s8 to some other devices means imitating a keyboard or mouse,  i checked the above fix in "deafult.prop" it have "persist.sys.usb.config=sec_charging" not "mtp" and i changed it to "sec_chariging,hid" is there any problem?

**Comment by ace2nutzer** [comment]
use persist.sys.usb.config=hid
U've also to add proper confguration in init.usb.configfs.rc for hid.

For instance here is an example:

```
# USB HID configuration
on property:sys.usb.config=hid && property:sys.usb.configfs=1
    write /sys/class/android_usb/android0/enable 0
    write /sys/kernel/config/usb_gadget/g1/idProduct 0x2d00
    write /sys/kernel/config/usb_gadget/g1/idVendor 0x18d1
    write /sys/class/android_usb/android0/functions ${sys.usb.config}
    write /sys/kernel/config/usb_gadget/g1/UDC ${sys.usb.controller}
    write /sys/class/android_usb/android0/enable 1
    setprop sys.usb.state ${sys.usb.config}
```
If it doesn't work, make sure on a booted Android that sys.usb.config is still =hid, if not set it and then connect the USB device.
If it doesn't help too, you may need to update the idProduct && idVendor. (search the web, or take it from any other kernel where hid is enabled and working)

**Comment by sploitpayt** [comment]
Is there any other way to contact you like telegram or something to make a proper conversation

**Comment by ace2nutzer** [comment]
No live-chat support.

**Comment by sploitpayt** [comment]
Ok, i checked s7 kernel which generate hidg0 in /dev and the ramdisk is not configured like you said. so is there any possibility to generate hidg0 in s8 Like s7

**Comment by ace2nutzer** [comment]
I've no idea, but you could compare defconfig S7 vs. S8, maybe you find what's missing or what should be disabled to make it working like S7.

**Comment by ace2nutzer** [comment]
Re-opened since there is still no fix.
